### PR TITLE
Convert "Invalid source root" errors to UserErrors

### PR DESCRIPTION
**Pull request by aeisenberg**
This is an internal change. No changelog note required.

### Merge / deployment checklist

- [x] Confirm this change is backwards compatible with existing workflows.
- [x] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [x] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

File: src/init.ts [hunk 1]
```diff
@@ -117,9 +117,10 @@ export async function runInit(
       }
     }
   } catch (e) {
-    // Handle the situation where init is called twice
-    // for the same database in the same job.
+    // Check if we need to re-throw the error as a UserError in order to avoid
+    // counting this error towards our internal error budget.
     if (
```
Comment (adityasharad) [inline src/init.ts:122]: Time to extract these conditions into their own `shouldReportAsUserError` helper function?

Comment (aeisenberg) [reply src/init.ts:122]: sure

**Comment by aeisenberg** [comment]
Rebasing in order to pick up a fix in integration tests from https://github.com/github/codeql-action/pull/1281.

File: src/init.ts [hunk 2]
```diff
@@ -144,6 +127,40 @@ export async function runInit(
...
+      `Is the "init" action called twice in the same job? ${e.message}`
+    );
+  } else if (
+    // Version of CodeQL CLI is incompatible with this version of the CodeQL Action
+    e.message?.includes("is not compatible with this CodeQL CLI") ||
+    // Expected source location for database creation does not exist
+    e.message?.includes("Invalid source root")
+  ) {
+    return new util.UserError(e.message);
+  } else {
```
Comment (adityasharad) [inline src/init.ts:159]: Similar, this can just be a final fall-through return statement.

**Review by adityasharad — DISMISSED** [review]
Looks good, minor suggestions only.

File: src/init.ts [hunk 3]
```diff
@@ -144,6 +127,42 @@ export async function runInit(
...
+ * Possibly convert this error into a UserError in order to avoid
+ * counting this error towards our internal error budget.
+ *
+ * @param e The error to possibly convert to a UserError.
+ *
+ * @returns A UserError if the error is a known error that can be
+ *         attributed to the user, otherwise the original error.
+ */
+function processError(e: any): Error {
+  if (!(e instanceof Error)) {
```
Comment (angelapwen) [inline src/init.ts:140]: Just curious: when would `e` not be an `instanceof Error` if it was sent from the `try/catch` block? Or is this just good practice in any error handling function?

File: src/init.ts [hunk 4]
```diff
@@ -144,6 +127,42 @@ export async function runInit(
...
+    return e;
+  }
+
+  if (
+    // Init action called twice
+    e.message?.includes("Refusing to create databases") &&
+    e.message.includes("exists and is not an empty directory.")
```
Comment (angelapwen) [inline src/init.ts:147]:
```suggestion
    e.message?.includes("exists and is not an empty directory.")
```

**Review by angelapwen — COMMENTED** [review]
Just a question/suggestion, not approving yet because it'll probably be dismissed by another commit anyway.

File: src/init.ts (hunk 3 above)
Comment (aeisenberg) [reply src/init.ts:140]: In javascript, you can "throw" anything. So you are not guaranteed that a catch clause is actually catching an `Error` instance. In practice, you should only throw errors. Typescript is just being pedantic here, and so is this code.

File: src/init.ts (hunk 4 above)
Comment (aeisenberg) [reply src/init.ts]: Oops. This commit is going to fail since the generated code is not up to date.

Comment (angelapwen) [reply src/init.ts]: Ah, I see, that makes sense!

**Review by angelapwen — APPROVED** [review]

**Pull request merged**